### PR TITLE
Switch to typescript

### PR DESCRIPTION
**Pull request by egoist**
- [ ] Add `ware.d.ts`
- [ ] Add `index.d.ts`
- [ ] Directly import `./src/index.ts` in Jest tests

### CODE REVIEW

**Comment by ulivz** [comment]
Shall we add tslint or prettier?

File: src/index.ts [hunk 1]
```diff
@@ -1,15 +1,42 @@
...
+
+interface StatCache {
+  [path: string]: fs.Stats
+}
+
 class Majo {
+  middlewares: Middleware[]
+  meta: any
+  baseDir: string
+  sourcePatterns: Glob
```
Comment (ulivz) [inline src/index.ts]: It's even better to add `private/public/protected` declaration at each field. I guess the `sourcePatterns` and `dotFiles` are private fields 😅

File: src/index.ts [hunk 2]
```diff
@@ -1,15 +1,42 @@
 import path from 'path'
 import fs from 'fs-extra'
 import globby from 'globby'
+// @ts-ignore
 import ware from 'ware'
 
+type Middleware = (ctx: Majo) => any
+type Glob = string | string[]
+type TransformFn = (contents: string) => Promise<string> | string
```
Comment (ulivz) [inline src/index.ts]: Shall we add support for returning `Buffer` and `stream`?

File: src/index.ts [hunk 3]
```diff
@@ -65,21 +92,21 @@ class Majo {
...
+  transform(relative: string, fn: TransformFn) {
     const contents = this.files[relative].contents.toString()
     const result = fn(contents)
-    if (!result.then) {
+    if (typeof result === 'string') {
       this.files[relative].contents = Buffer.from(result)
       return
     }
-    return result.then(newContents => {
+    return result.then((newContents: string) => {
```
Comment (ulivz) [inline src/index.ts]: I think it's not making sense, when user write a transform function that returns a number, the error message will be `result.then is not a function`. the user will be very confused so he must to read the source code.

**Comment by egoist** [comment]
@ulivz Yeah, but let's fix existing issues first 😄

**Pull request closed without merging**